### ISSUE TITLE
ais etl error: ConnectionResetError: [Errno 104] Connection reset by peer

### ISSUE THREAD

**Issue by yingca1** [NVIDIA/aistore#146]
```bash
Exception occurred during processing of request from ('10.232.8.1', 64979)
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/socketserver.py", line 683, in process_request_thread
    self.finish_request(request, client_address)
  File "/usr/local/lib/python3.10/socketserver.py", line 360, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/usr/local/lib/python3.10/socketserver.py", line 747, in __init__
    self.handle()
  File "/usr/local/lib/python3.10/http/server.py", line 433, in handle
    self.handle_one_request()
  File "/usr/local/lib/python3.10/http/server.py", line 421, in handle_one_request
    method()
  File "/code/server.py", line 101, in do_GET
    self.wfile.write(processed_data)
  File "/usr/local/lib/python3.10/socketserver.py", line 826, in write
    self._sock.sendall(b)
ConnectionResetError: [Errno 104] Connection reset by peer
```

I often encounter the above errors during the ais etl testing process, probably at a rate of 10% to 15%. 

Where should I look to troubleshoot this issue?

**Comment by gaikwadabhishek** [comment]
Hello @yingca1!

Could you kindly explain the operation you are currently performing and some insights into the type and parameters used to initialize ETL?

I have encountered similar errors in the past when switching between networks or using a VPN.

To assist with debugging, you can follow these steps:

1. Print all the logs on the container pod using the following command:
```
$ ais etl view-logs <etl-name>
```

By reviewing the logs, we can gain a better understanding of the issue and work towards resolving it effectively. Thank you!

**Comment by yingca1** [comment]
I use bucket ETL operations.
``` bash
ais etl bucket transformer-etl ais://dataset1 ais://out1
```

**Comment by yingca1** [comment]
``` bash
ais config cluster timeout.send_file_time=10m
ais config cluster timeout.max_keepalive=30s
ais config cluster client.client_timeout=2m
```
after this config set, we have 95% success, but still have some failed case with the same error

I will check the ETL task consumed time.

**Comment by alex-aizman** [comment]
To put `send_file_timeout` in perspective, sending 10GB payload will take about 14m over a 100Mbps network including overhead but not including disk reading times. On the other hand, if your network is 100Gbps than it's less then one second. And so here we go - initial config used for deployment cannot be a simple copy of the default template we provide - it won't work or rather, won't always work.

We could, of course, have `send_file_time = 10h` or remove it altogether... but there's a price: missing or masking a bigger potential problem, etc.

**Comment by yingca1** [comment]
update:

There is still an issue, but I've observed that it happens less often on bigger clusters. A 150-node aistore cluster has considerably fewer occurrences compared to a 30-node cluster.

**Comment by gaikwadabhishek** [comment]
@yingca1 , we are looking into it. We are trying out different web servers to see if other web servers are able to handle the load. We have experimented with [flask + gunicorn](https://github.com/NVIDIA/ais-etl/tree/master/transformers/keras_preprocess), it was slightly better but still seeing some issues on very high workloads.

**Comment by gaikwadabhishek** [comment]
Hey @yingca1,
I have some updates, I've been working on this issue and trying out different things for running transformers - https://github.com/NVIDIA/ais-etl/tree/master/bench. 

```
 if arg_type == "url":
            processed_data = wd_etl(query_path)
        else:
            input_bytes = requests.get(query_path).content
            processed_data = wd_etl(input_bytes)
```
Inside this wd_etl, we are making a GET call to the object, and in case of offline transforms multiple get calls are made simultaneously. These calls are reading from files on target and if the machine is not able to handle the amount of requests, some of the initial calls are closed to accommodate new ones, which might lead to that issue. 

Can you try the whole thing without Webdataset once? Try to see examples under https://github.com/NVIDIA/ais-etl/tree/master/transformers we have different transformers running different webservers and all the necessary files for creating it will be available there. 

We are still working on creating a one-stop solution for these kinds of problems. I will get back to you once we have a fully working solution.

**Comment by gaikwadabhishek** [comment]
Hello there, @yingca1!

I've recently implemented certain enhancements to a selection of transformers, resulting in improved speed and optimized request handling. I kindly request you to pattern match your `server.py` to our [hello world transformer](https://github.com/NVIDIA/ais-etl/tree/master/transformers/hello_world) or [keras transformer's](https://github.com/NVIDIA/ais-etl/tree/master/transformers/keras_preprocess) server. During my benchmarking attempts, I observed that the servers and requests were operating without throttling, leading to an overall enhancement in performance and faster execution.

Feel free to reach out if you require any assistance. Kindly keep me informed about the outcome once you've had the chance to implement these changes. 

-Abhishek
